pkg/trie/trie.go

```diff
@@ -1,1518 +1,81 @@
-func (t *Trie) RootNode() *Node {
-	copySettings := node.DefaultCopySettings
-	copySettings.CopyMerkleValue = true
-	return t.root.Copy(copySettings)
+type KVStore interface {
+	Get(key []byte) []byte
+	Put(key, value []byte) error
+	Delete(key []byte) error
 }
 
-// MustHash returns the hashed root of the trie.
-// It panics if it fails to hash the root node.
-func (t *Trie) MustHash() common.Hash {
-	h, err := t.Hash()
-	if err != nil {
-		panic(err)
-	}
-
-	return h
+type TrieIterator interface {
+	Entries() (keyValueMap map[string][]byte)
+	NextKey(key []byte) []byte
 }
 
-// Hash returns the hashed root of the trie.
-func (t *Trie) Hash() (rootHash common.Hash, err error) {
-	if t.root == nil {
-		return EmptyHash, nil
-	}
-
-	merkleValue, err := t.root.CalculateRootMerkleValue()
-	if err != nil {
-		return rootHash, err
-	}
-	copy(rootHash[:], merkleValue)
-	return rootHash, nil
+type PrefixTrie interface {
+	GetKeysWithPrefix(prefix []byte) (keysLE [][]byte)
+	ClearPrefix(prefix []byte) (err error)
+	ClearPrefixLimit(prefix []byte, limit uint32) (
+		deleted uint32, allDeleted bool, err error)
 }
 
-// Entries returns all the key-value pairs in the trie as a map of keys to values
-// where the keys are encoded in Little Endian.
-func (t *Trie) Entries() (keyValueMap map[string][]byte) {
-	keyValueMap = make(map[string][]byte)
-	t.buildEntriesMap(t.root, nil, keyValueMap)
-	return keyValueMap
+type TrieDeltas interface {
+	GetChangedNodeHashes() (inserted, deleted map[common.Hash]struct{}, err error)
+	handleTrackedDeltas(success bool, pendingDeltas tracking.Getter)
 }
 
-func (t *Trie) buildEntriesMap(currentNode *Node, prefix []byte, kv map[string][]byte) {
-	if currentNode == nil {
-		return
-	}
-
-	// Leaf
-	if currentNode.Kind() == node.Leaf {
-		key := currentNode.PartialKey
-		fullKeyNibbles := concatenateSlices(prefix, key)
-		keyLE := codec.NibblesToKeyLE(fullKeyNibbles)
-		kv[string(keyLE)] = t.Get(keyLE)
-		return
-	}
-
-	// Branch
-	branch := currentNode
-	if branch.StorageValue != nil {
-		fullKeyNibbles := concatenateSlices(prefix, branch.PartialKey)
-		keyLE := codec.NibblesToKeyLE(fullKeyNibbles)
-		kv[string(keyLE)] = t.Get(keyLE)
-	}
-
-	for i, child := range branch.Children {
-		childPrefix := concatenateSlices(prefix, branch.PartialKey, intToByteSlice(i))
-		t.buildEntriesMap(child, childPrefix, kv)
-	}
+type DBBackedTrie interface {
+	Load(db db.DBGetter, rootHash common.Hash) error
+	WriteDirty(db NewBatcher) error
+	writeDirtyNode(db db.DBPutter, n *Node) (err error)
 }
 
-// NextKey returns the next key in the trie in lexicographic order.
-// It returns nil if no next key is found.
-func (t *Trie) NextKey(keyLE []byte) (nextKeyLE []byte) {
-	prefix := []byte(nil)
-	key := codec.KeyLEToNibbles(keyLE)
-
-	nextKey := findNextKey(t.root, prefix, key)
-	if nextKey == nil {
-		return nil
-	}
-
-	nextKeyLE = codec.NibblesToKeyLE(nextKey)
-	return nextKeyLE
+type Printable interface {
+	String() string
```

Comment:
I would suggest to use `fmt.Stringer `

Reply:
Great! I didn't know it, I'm going to change it